refactoring(circuits) - cleanup circuits package, comment code and split into utilities vs exported functions (#832)
privacy-scaling-explorations/maci: website/versioned_docs/version-v1.x/circuits.md

Reviewer: Can you check the styles on diagrams? I noticed some of them don't fit the text.

Author: As in you cannot read the text?

Reviewer: I can but it doesn't look pretty. It's like text overlapping the rectangle. So I suggest to make rectangles bigger so they fit all the text.

# Circuits

MACI has three main zk-SNARK [circuits](https://github.com/privacy-scaling-explorations/maci/tree/dev/circuits):

> 1. [`ProcessMessages.circom`](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/processMessages.circom), which takes a batch of encrypted messages, decrypts them, and generates a proof that the coordinator's local processing was performed correctly.

Reviewer 2: Nitpick but I think this should be `ProcessMessages` (circuit name, PascalCase) or `processMessages.circom` (file name, camelCase)

Naming convention question: _should_ Circom files be camelCase?

Reviewer 2: Might just be easiest to update the filenames to be PascalCase to match the circuit names?

Author: Not sure about conventions, here is how [circomlib](https://github.com/iden3/circomlib/tree/d5ed1c3ce4ca137a6b3ca48bec4ac12c1b38957a/circuits) names them (all lower it seems)

2. [`TallyVotes.circom`](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/tallyVotes.circom), which counts votes from users' ballots, batch by batch.
3. [`Subsidy.circom`](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/subsidy.circom), which implements [pairwise subsidy](https://hackmd.io/@chaosma/H1_9xmT2K).

> Each circuit is parameterised and it is important to set the right parameters that matches your use case. For example, if you want to support up to 3125 messages, the message tree depth parameter should be set to `5` (as $5^5 = 3125$).

Reviewer 2: ```suggestion
The rest of the circuits are utilities templates that are required for the main circuits to work correctly. These include utilities such as float math, conversion of private keys, and Poseidon hashing/encryption.

Each circuit is parameterised and it is important to set the right parameters that matches your use case. For example, if you want to support up to 3125 messages, the message tree depth parameter should be set to `5` (as $5^5 = 3125$).
```
Copying over blurb from the README: https://github.com/privacy-scaling-explorations/maci/tree/dev/circuits

## Background

### zk-SNARKs

> MACI uses zk-SNARKs to prove that the coordinator has correctly processed a batch of messages and that the votes have been tallied correctly. zk-SNARKs are a type of zero-knowledge proof which allows a prover to prove that they know a secret without revealing the secret itself. In MACI, the prover is the coordinator. Users, can at any point after the proof generation, verify that the process was done correctly.

Reviewer 2: ```suggestion
zk-SNARKs are a type of zero-knowledge proof which allows a "prover" to prove to a "verifier" that they know a secret without revealing the secret itself. In MACI, the prover is the coordinator. MACI uses zk-SNARKs to prove that the coordinator has correctly processed the batches of messages and that all votes have been tallied correctly. A smart contract acts as the verifier to check the proof from the coordinator. Users can also verify that the process was done correctly at any point after the proof generation.
```

### Circom

MACI's circuits are written using [Circom](https://docs.circom.io/), a domain-specific language (DSL) used to write zk-SNARK circuits. Circom syntax resembles JavaScript, and it currently is one of the most popular DSL in use by zk developers. Please refer to their documentation to learn more about the language.

### Proving system

> MACI uses [GROTH16](https://eprint.iacr.org/2016/260.pdf) as its proving system. GROTH16 is a zk-SNARK proving system that allows for the generation of proofs that are small and fast to verify.

Reviewer 2: ```suggestion
MACI uses [Groth16](https://eprint.iacr.org/2016/260.pdf) as its proving system. Groth16 is a zk-SNARK proving system that allows for the generation of proofs that are small and fast to verify.
```
From the use I've seen...

> ## How are the circuits used

Reviewer 2: ```suggestion
## How are circuits are used
```

> ## How are the circuits used

Reviewer 2: ```suggestion
## How are the circuits used?
```

> The circuits are used by the coordinator (the prover) to prove that they have correctly processed a batch of messages and tallied the votes correctly. This happens after a Poll has completed, and the coordinator has merged the state and message trees. The coordinator then generates a proof for each batch of messages, and submits them to the contract. The contract then verifies the proofs and updates the commitments on chain.

Reviewer 2: > correctly processed a batch of messages and tallied the votes correctly

General terminology question - can/should "messages" & "votes" be used interchangeably? Or are votes a specific type of message (i.e. messages can be key switch commands or vote commands)?

Author: Correct, votes are a specific type of messages. A message could simply be a topup message requesting to topup a participant's voice credits

> > Please note that in order to function, the circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.

Reviewer 2: ```suggestion
Please note that in order to function, the circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.
```
Why the special quote formatting here?

Author: Like an important note or something

> > Please note that in order to function, the circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.

Reviewer 2: ```suggestion
The circuit in charge of processing the messages requires the coordinator's private key, hence a proof for this circuit can only be generated by the coordinator. The private key is needed in order to generate the ECDH shared key used to decrypt the messages.
```

Reviewer 2: This paragraph might make the most sense under the "Message processing" section

> ## How do the Circuits fit in a voting round

Reviewer 2: ```suggestion
## How do the Circuits fit in a voting round?
```

> ![flow](/img/circuits/processingAfterPollEnds.svg)

Reviewer 2: These flow charts are awesome! 🤩 

I have some improvement suggestions but I think that's best saved for a future PR

Author: Feel free to edit the file yourself! included the source in static/img/circuits which you can load into [excalidraw](https://excalidraw.com/)

> ## How the circuits work

Reviewer 2: ```suggestion
## How do the circuits work?
```
Keeping the consistency of question format

> ### Message processing (processMessages)

Reviewer 2: For "processMessages", maybe `ProcessMessages` (circuit name in code) or processMessages.circom (file) or link to the circuit to make clear that's the circuit name?

> This circuit allows the coordinator to prove that they have correctly processed each message in reverse order, in a consecutive batch of 5 ^ msgBatchDepth messages to the respective state leaf within the state tree. Coordinators would use this circuit to prove correct execution at the end of each Poll. The pre-requisites for this are:

Reviewer 2: I think worth explaining why this happens in batches (because of gas limit, correct?)

Author: Well AFAIU we do it in batches because of circuits' constraints. If we were to process all messages in one go, the constraints would be so many that it would take forever and the circuit's artifacts (zkeys, r1cs, etc) would be MASSIVE

> This circuit allows the coordinator to prove that they have correctly processed each message in reverse order, in a consecutive batch of 5 ^ msgBatchDepth messages to the respective state leaf within the state tree. Coordinators would use this circuit to prove correct execution at the end of each Poll. The pre-requisites for this are:

Reviewer 2: ```suggestion
This circuit allows the coordinator to prove that they have correctly processed each message in reverse order, in a consecutive batch of 5 ^ msgBatchDepth messages to the respective state leaf within the state tree. Coordinators would use this circuit to prove correct execution at the end of each Poll.

The pre-requisites for this circuit are:
```

Reviewer 2: Added your explanation from one of my comments, which I thought was helpful 😄

- the related Poll has ended
- the state tree has been merged
- the message tree has been merged

![ProcessMessages](/img/circuits/processMessages.svg)

#### Parameters

| #   | Parameter                | Description                                          |
| --- | ------------------------ | ---------------------------------------------------- |
| 0   | State tree depth         | Allows $(5^{n})$ signups.                            |
| 1   | Message tree depth       | Allows $(5^{n})$ votes or key-change messages.       |
| 2   | Message batch tree depth | Allows $(5^{n})$ messages to be processed per batch. |
| 3   | Vote option tree depth   | Allows $(5^{n})$ vote options.                       |

#### Inputs

> | Input signal                     | Description                                                                             |

Reviewer 2: This table is great, thanks for doing this

Author: oh no can't take props for that, it's a copy pasta from here: https://maci.pse.dev/docs/spec 🙂

The implementation for this is in the `QuinBatchLeavesExists` circuit in `https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/trees/incrementalQuinTree.circom`.

This method requires fewer circuit constraints than if we verified a Merkle proof for each leaf.

#### Statements that the circuit proves

1. That the prover knows the preimage to `inputHash` (see above)
2. That the prover knows the preimage to `currentSbCommitment` (that is, the state root, ballot root, and `currentSbSalt`)
3. That `maxVoteOptions <= (5 ^ voteOptionTreeDepth)`
4. That `numSignUps <= (5 ^ stateTreeDepth)`
5. That `coordPubKey` is correctly derived from `coordPrivKey`
6. That `coordPubKey` is the preimage to the Poseidon hash of `coordPubKey` (provided by the contract)
7. That each message in `msgs` exists in the message tree
8. That after decrypting and applying each message, in reverse order, to the corresponding state and ballot leaves, the new state root, new ballot root, and `newSbSalt` are the preimage to `newSbCommitment`

> ### Tally Votes

Reviewer 2: Explanation would be nice here (but as I'm reviewing I'm thinking a lot of these documentation changes could be done in future PRs)

> ### Tally Votes

Reviewer 2: ```suggestion
### Tally Votes (`TallyVotes`)
```

Reviewer 2: (or whatever the same convention is we decide on above for `ProcessMessages`)

   - This is the hash of the Merkle root $r_r$ of the votes and a salt $r_s$, computed as $poseidon_2([r_r, r_s])$
2. $tc_t$: a commitment to the total spent voice credits
   - This is the hash of the total spent voice credits $t_c$ and a salt $t_s$, computed as $poseidon_2([t_c, t_s])$
3. $tc_p$: a commitment to the spent voice credits per vote option
   - This is the hash of the Merkle root of the spent voice credits per vote option $p_v$ and a salt $p_s$, computed as $poseidon_2([p_v, p_s])$

The tally commitment is computed as such:

$poseidon_3([tc_r, tc_t, tc_p])$

#### Statements that the circuit proves

1. That the coordinator knows the preimage of `sbCommitment`
2. That the coordinator knows the preimage of `inputHash`
3. That `batchStartIndex` is less than or equal to `numSignUps`
4. That each ballot in `ballots` is in a member of the ballot tree with the Merkle root `ballotRoot` at indices `batchStartIndex` to `batchStartIndex + (5 ** intStateTreeDepth)`
5. That each set of votes (`votes[i]`) has the Merkle root $blt_r$ whose value equals `ballots[i][1]`
6. That the tally is valid, which is:
   - That the sum of votes per vote option is correct

> ### Subsisdy (Optional)

Reviewer 2: ```suggestion
### Subsidy (`Subsidy`)
```

> The subsidy circuit is used to implement pairwise subsidy. This is a technique that can be used to detect voters collusion. It currently is not optimized for production and the team will work on a more efficient implementation in the future.

Reviewer 2: ```suggestion
This circuit is an optional feature - it's not required for MACI to function. The subsidy circuit is used to implement pairwise subsidy. This is a technique that can be used to detect voters collusion. It currently is not optimized for production and the team will work on a more efficient implementation in the future.
```

# Circuits

> MACI has three zk-SNARK [circuits](https://github.com/privacy-scaling-explorations/maci/tree/dev/circuits):

Reviewer 2: ```suggestion
MACI has three main zk-SNARK [circuits](https://github.com/privacy-scaling-explorations/maci/tree/dev/circuits):
```
Added "main" - MACI has way more than 3 circuits (right?)

> 1. [`ProcessMessages.circom`](https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/processMessages.circom), which takes a batch of messages, and updates the

Reviewer 2: > which takes a batch of messages, and updates the state and ballot trees according to the contents of said messages.

This confuses me - how does a circuit "update the state and ballot trees"? It's not actually updating state anywhere is it? How can a circuit alter state elsewhere?

Author: I see how this can be misleading, I will rephrase! basically I think what they meant to say there is that the processMessages circuit will try to decrypt the messages, and based on the content of the message, update within itself the trees, to generate a proof that the processing was done correctly offchain! In other words, the circuit takes a final state, an initial state, and the leaves (messages and user signups) - it process these messages via the different state transitions to finally check that the expected state is correct

Reviewer 2: > I see how this can be misleading, I will rephrase! basically I think what they meant to say there is that the processMessages circuit will try to decrypt the messages, and based on the content of the message, update within itself the trees, to generate a proof that the processing was done correctly offchain! In other words, the circuit takes a final state, an initial state, and the leaves (messages and user signups) - it process these messages via the different state transitions to finally check that the expected state is correct

Ah, great explanation, thanks!
